File: python/백준/재귀/1074 Z.py
import sys
input = sys.stdin.readline

N, r, c = map(int, input().split())

count = 0  # 방문 순서

def search(x, y, size):
    global count
    # 크기가 1x1이 되면 값 출력
    if size == 1:
        if x == r and y == c:
            print(count)
            exit()  # 프로그램 종료
        count += 1
        return

    # 영역을 4분할
    new_size = size // 2

    # 각 영역의 시작점에서 재귀 호출 (Z 모양 순서)
    if r < x + new_size and c < y + new_size:  # 왼쪽 위
        search(x, y, new_size)
    elif r < x + new_size and c >= y + new_size:  # 오른쪽 위
        count += new_size * new_size  # 이전 영역 크기만큼 count 증가
        search(x, y + new_size, new_size)
    elif r >= x + new_size and c < y + new_size:  # 왼쪽 아래
        count += 2 * new_size * new_size  # 이전 영역 크기만큼 count 증가
        search(x + new_size, y, new_size)
    else:  # 오른쪽 아래
        count += 3 * new_size * new_size  # 이전 영역 크기만큼 count 증가
        search(x + new_size, y + new_size, new_size)

# 탐색 시작
search(0, 0, 2 ** N)

File: cpp/백준/재귀/1074 Z.cpp
#include <iostream>
using namespace std;

int N, r, c;
long long count_ = 0;

void searchZ(int x, int y, long long size) {
    if (size == 1) {
        if (x == r && y == c) {
            cout << count_ << "\n";
            exit(0);
        }
        count_ += 1;
        return;
    }

    long long new_size = size / 2;

    if (r < x + new_size && c < y + new_size) {
        searchZ(x, y, new_size);
    } else if (r < x + new_size && c >= y + new_size) {
        count_ += new_size * new_size;
        searchZ(x, y + new_size, new_size);
    } else if (r >= x + new_size && c < y + new_size) {
        count_ += 2 * new_size * new_size;
        searchZ(x + new_size, y, new_size);
    } else {
        count_ += 3 * new_size * new_size;
        searchZ(x + new_size, y + new_size, new_size);
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> r >> c;
    searchZ(0, 0, 1LL << N);
    return 0;
}
